File: backend/tasks/agents/atlas_analyst/prompt_loader.py
"""Simple prompt loader with template composition and variable substitution."""

import os
import re
import yaml


class PromptLoader:
    """Loads prompts from YAML and resolves template references."""

    def __init__(self, prompts_file: str = None):
        """Initialize prompt loader.

        Args:
            prompts_file: Path to YAML file. Defaults to prompts.yaml in same directory.
        """
        if prompts_file is None:
            current_dir = os.path.dirname(__file__)
            prompts_file = os.path.join(current_dir, "prompts.yaml")

        with open(prompts_file, 'r') as f:
            data = yaml.safe_load(f)

        self.templates = data.get('templates', {})
        self.prompts = data.get('prompts', {})

    def get(self, prompt_id: str, **variables) -> str:
        """Get prompt by ID with variable substitution.

        Args:
            prompt_id: Prompt identifier (e.g., 'system', 'user')
            **variables: Variables to substitute in the prompt

        Returns:
            Fully composed and substituted prompt string

        Example:
            >>> loader = PromptLoader()
            >>> prompt = loader.get(
            ...     'system',
            ...     schema='...',
            ...     context='...',
            ...     connection_id='123',
            ...     max_steps=10
            ... )
        """
        # Get the prompt template
        prompt = self._get_nested(self.prompts, prompt_id)
        if prompt is None:
            raise ValueError(f"Prompt '{prompt_id}' not found")

        # Resolve template references ({template_name} or {template.path})
        prompt = self._resolve_templates(prompt)

        # Substitute variables
        prompt = prompt.format(**variables)

        return prompt

    def _get_nested(self, data: dict, path: str):
        """Get nested value from dict using dot notation path.

        Args:
            data: Dictionary to search
            path: Dot-separated path (e.g., 'system.question', 'context.schema')

        Returns:
            Value at path or None if not found
        """
        keys = path.split('.')
        current = data

        for key in keys:
            if isinstance(current, dict) and key in current:
                current = current[key]
            else:
                return None

        return current

    def _resolve_templates(self, text: str) -> str:
        """Resolve template references in text.

        Handles:
        - {path.to.template} style nested references (with dots)
        - {template_name} style simple references (single word, if exists in templates)

        Args:
            text: Text containing template references

        Returns:
            Text with all template references resolved
        """
        # Pattern to match {path.to.template} (nested with dots)
        nested_pattern = re.compile(r'\{([\w]+\.[\w.]+)\}')

        # Pattern to match simple {template_name} (single word)
        simple_pattern = re.compile(r'\{(\w+)\}')

        # Keep resolving until no more references found (handles nested templates)
        max_iterations = 10  # Prevent infinite loops
        for _ in range(max_iterations):
            made_replacement = False

            # Resolve {path.to.template} style
            for template_path in nested_pattern.findall(text):
                template_content = self._get_nested(self.templates, template_path)
                if template_content is None:
                    raise ValueError(f"Template '{template_path}' not found")
                text = text.replace(f'{{{template_path}}}', template_content)
                made_replacement = True

            # Resolve simple {template_name} style (only if template exists)
            for template_name in simple_pattern.findall(text):
                if template_name in self.templates:
                    template_content = self.templates[template_name]
                    if isinstance(template_content, str):
                        text = text.replace(f'{{{template_name}}}', template_content)
                        made_replacement = True

            if not made_replacement:
                break

        return text


# Global instance for easy import
_loader = None


def get_prompt(prompt_id: str, **variables) -> str:
    """Convenience function to get a prompt using global loader.

    Args:
        prompt_id: Prompt identifier
        **variables: Variables to substitute

    Returns:
        Composed and substituted prompt
    """
    global _loader
    if _loader is None:
        _loader = PromptLoader()
    return _loader.get(prompt_id, **variables)

File: frontend/agents/src/analyst/prompt-loader.ts
import { readFileSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';
import yaml from 'js-yaml';

// Resolve this module's directory in both ESM and CJS contexts. ts-jest runs the
// orchestrator project as ESM (import.meta.url) and the main project as CJS (__dirname).
// We reference import.meta inside a try block so the CJS branch never evaluates it.
function moduleDir(): string {
  // CJS path first — when __dirname exists, import.meta is a syntax error in some runtimes.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const maybeDirname = (globalThis as any).__dirname ?? (typeof __dirname !== 'undefined' ? __dirname : undefined);
  if (typeof maybeDirname === 'string') return maybeDirname;
  // ESM path — import.meta.url is the URL of this module.
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const metaUrl = (import.meta as any)?.url as string | undefined;
  if (metaUrl) return dirname(fileURLToPath(metaUrl));
  throw new Error('Cannot resolve module directory (neither __dirname nor import.meta.url available)');
}

/**
 * Port of `backend/tasks/agents/analyst/prompt_loader.py`.
 *
 * Loads prompts.yaml, resolves nested `{template}` and `{path.to.template}` references,
 * and substitutes `{variable}` placeholders. Mirrors Python `str.format(**vars)` semantics:
 * unsubstituted `{var}` placeholders raise.
 */

export const HIDDEN_SKILLS = new Set(['navigation_restricted', 'navigation_unrestricted']);

interface PromptsFile {
  templates: Record<string, unknown>;
  prompts: Record<string, unknown>;
}

export class PromptLoader {
  readonly templates: Record<string, unknown>;
  readonly prompts: Record<string, unknown>;

  constructor(promptsFile?: string) {
    const path = promptsFile ?? join(moduleDir(), 'prompts.yaml');
    const raw = readFileSync(path, 'utf8');
    const data = yaml.load(raw) as PromptsFile;
    this.templates = data.templates ?? {};
    this.prompts = data.prompts ?? {};
  }

  get(promptId: string, variables: Record<string, unknown> = {}): string {
    const prompt = this.getNested(this.prompts, promptId);
    if (typeof prompt !== 'string') {
      throw new Error(`Prompt '${promptId}' not found or is not a string`);
    }
    const resolved = this.resolveTemplates(prompt);
    return this.formatVariables(resolved, variables);
  }

  listSkills(skipHidden = false): Record<string, string> {
    const skills: Record<string, string> = {};
    for (const [key, value] of Object.entries(this.templates)) {
      if (key.startsWith('skill_') && typeof value === 'object' && value !== null) {
        const name = key.slice('skill_'.length);
        if (skipHidden && HIDDEN_SKILLS.has(name)) continue;
        skills[name] = (value as { description?: string }).description ?? '';
      }
    }
    return skills;
  }

  getSkill(name: string): string | null {
    const key = `skill_${name}`;
    const template = this.templates[key];
    if (!template || typeof template !== 'object') return null;
    const content = (template as { content?: string }).content ?? '';
    if (!content) return null;
    return this.resolveTemplates(content);
  }

  private getNested(data: Record<string, unknown>, path: string): unknown {
    let current: unknown = data;
    for (const key of path.split('.')) {
      if (typeof current === 'object' && current !== null && key in current) {
        current = (current as Record<string, unknown>)[key];
      } else {
        return null;
      }
    }
    return current;
  }

  private resolveTemplates(text: string): string {
    const nestedPattern = /\{(\w+(?:\.\w+)+)\}/g;
    const simplePattern = /\{(\w+)\}/g;

    let current = text;
    for (let i = 0; i < 10; i++) {
      let madeReplacement = false;

      // Resolve nested {a.b.c} first
      current = current.replace(nestedPattern, (match, templatePath: string) => {
        const content = this.getNested(this.templates, templatePath);
        if (content === null) {
          throw new Error(`Template '${templatePath}' not found`);
        }
        madeReplacement = true;
        return String(content);
      });

      // Then resolve simple {name} only if it exists in templates
      current = current.replace(simplePattern, (match, name: string) => {
        if (name in this.templates) {
          const content = this.templates[name];
          if (typeof content === 'string') {
            madeReplacement = true;
            return content;
          }
        }
        return match;
      });

      if (!madeReplacement) break;
    }
    return current;
  }

  private formatVariables(text: string, variables: Record<string, unknown>): string {
    // Python str.format substitutes {name}; we leave non-matching {…} as-is rather than throw.
    return text.replace(/\{(\w+)\}/g, (match, key: string) => {
      if (key in variables) {
        const v = variables[key];
        return v == null ? '' : String(v);
      }
      return match;
    });
  }
}

let cachedLoader: PromptLoader | null = null;
function getLoader(): PromptLoader {
  if (!cachedLoader) cachedLoader = new PromptLoader();
  return cachedLoader;
}

export function getPrompt(promptId: string, variables: Record<string, unknown> = {}): string {
  return getLoader().get(promptId, variables);
}

export function getSkill(name: string): string | null {
  return getLoader().getSkill(name);
}

export function listSkills(skipHidden = false): Record<string, string> {
  return getLoader().listSkills(skipHidden);
}
